### PR TITLE
Fix World#isChunkGenerated calls

### PR DESCRIPTION
**Pull request by Spottedleaf**
Optimize World#loadChunk() too
This patch also adds a chunk status cache on region files (note that
its only purpose is to cache the status on DISK)

Note that I still need to review this with a plugin, however debug reports show the concept works.
This fixes upstream issue https://hub.spigotmc.org/jira/browse/SPIGOT-4862

### CODE REVIEW

File: Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+-        if (chunk instanceof ProtoChunkExtension) {
+-            // We then cycle through again to get the full chunk immediately, rather than after the ticket addition
+-            chunk = world.getChunkProvider().getChunkAt(x, z, ChunkStatus.FULL, true);
+-        }
+-
+-        if (chunk instanceof net.minecraft.server.Chunk) {
+-            world.getChunkProvider().addTicket(TicketType.PLUGIN, new ChunkCoordIntPair(x, z), 1, Unit.INSTANCE);
+-            return true;
++        // Paper - Optimize this method
++        if (generate && !isChunkGenerated(x, z)) {
```
Comment (blablubbabc) [inline Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch]: Was probably meant to be
```suggestion
+        if (!generate && !isChunkGenerated(x, z)) {
```
Same below

Comment (Spottedleaf) [reply Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch]: I'll get the other one soon

File: Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+-        // If generate = false, but the chunk already exists, we will get this back.
+-        if (chunk instanceof ProtoChunkExtension) {
+-            // We then cycle through again to get the full chunk immediately, rather than after the ticket addition
+-            chunk = world.getChunkProvider().getChunkAt(x, z, ChunkStatus.FULL, true);
+-        }
+-
+-        if (chunk instanceof net.minecraft.server.Chunk) {
+-            return ((net.minecraft.server.Chunk)chunk).bukkitChunk;
++        // Note: Copied from loadChunk()
++        if (gen && !isChunkGenerated(x, z)) {
```
Comment (blablubbabc) [inline Spigot-Server-Patches/0403-Fix-World-isChunkGenerated-calls.patch]:
```suggestion
+        if (!gen && !isChunkGenerated(x, z)) {
```

**Comment by Spottedleaf** [comment]
Ready for review, passed plugin testing

Note that the ticket addition will cause chunks to generate around the target chunk if it is already loaded. There's nothing I can do about this but raise the added ticket's level; however that changes behavior for the chunk - it wont be able to tick. 
Plugins wishing to avoid this behavior can use our async chunk api, which does not add a plugin ticket, and does include options for generation (will still have that side-effect of not being able to tick, although that's already standard behavior for getChunkAt calls in 1.14)

**Pull request merged**